### ISSUE TITLE
Has anyone looked at loading parquet files?

### ISSUE THREAD

**Issue by cottrell** [OpenRefine/OpenRefine#1929]
Not too familiar with refine but could have a look at adding this if someone gives me a starting point.

**Comment by wetneb** [comment]
That is a good idea! Here are the main steps:
* add a dependency to a Java Parquet library (in `main/pom.xml`)
* write a `ParquetImporter` which implements `ImportingParser`: you will probably want to extend an existing class such as `TabularImportingParserBase` for that, using the classes from the Parquet library
* write an UI for the importing options in `./main/webapp/modules/core/scripts/index/parser-interfaces/`
* register the importer and its UI in `main/webapp/modules/core/MOD-INF/controller.js`

This can be also implemented as an extension (see the existing extensions for examples of how to set that up) but unless the Parquet library is very heavy or pulls in a lot of dependencies, I think this could fit in the core software easily.

**Comment by thadguidry** [comment]
@cottrell Both @jackyq2015 and myself only did some minor bikeshedding of ideas and Parquet was a format that was deemed potentially useful to us.  But we had planned to incorporate support for it by using Apache Beam.  See #1433 and #1468 as well as our **Performance Improvements** project page https://github.com/OpenRefine/OpenRefine/projects/1

**Comment by wetneb** [comment]
I plan to do this as part of the migration to Spark.

**Comment by wetneb** [comment]
This is no longer on my radar since I did not end up using Parquet as a storage format for project files. The main rationale behind this is that it is much simpler to use a text-based format, which can easily be implemented in all executors. We could still investigate if it is worth switching to Parquet though! (in a backwards-compatible way)

**Comment by dokutoshi** [comment]
I believe from a competitive view and usability, parquet with pyarrow engine is important.  A 81 million dataset in CSV format timed out after nearly 12 hours.  Concurrently, I used VSCode's new Data Wrangler and loaded the 81 million records (in CSV and Paraquet/Pyarrow) in less than 5 minutes with full statistics automatically populated 10 minutes later.  

Was quite disappointed with OpenRefine's responsive to a large CSV file.

**Comment by wetneb** [comment]
@dokutoshi I assume you tried OpenRefine 3?

The upcoming 4.0 version is designed to address this issue. The 4.0-alpha1 version is quite old but I plan to publish a new one soon and will be looking for testers for it. It would be amazing if you could be one, so I can work on making OpenRefine work for your use case. If you are interested, please write an email to antonin@delpeuch.eu.

Alternatively, if you want to try it out immediately, you could just run OpenRefine from our `4.0` branch.

**Comment by dokutoshi** [comment]
Let me give it a try if suitable for Mac Ventura Apple Silicon.

**Comment by simonaubertbd** [comment]
Hello. This would be a nice feature, Parquet is being more and more supported by tools and OpenRefine is almost an exception right now.

Best regards,

Simon

**Comment by thadguidry** [comment]
@simonaubertbd do you have needs to actually import a Parquet file into OpenRefine?  And if so, do you think it will be likely that you will need to wrangle the data in OpenRefine rather then using other tools designed well for Parquet wrangling and analysis?

**Comment by simonaubertbd** [comment]
Hello @thadguidry  and thanks for your answer. To be honest with you : yes there are other tools that can do the job but there are also other tools that can do when OpenRefine do and since I'm a partner of many software companies, I can find something else ;) However, it's not exactly for me, it's more for the OpenRefine lambda user who doesn't have the tools I have. Isn't your purpose to have the best software and good interoperability? Even LibreOffice will support Parquet (already in Beta). And Parquet becomes a standard in the data community, I mean almost all products support or will support it : Qlik, Alteryx in the next release, etc, etc- This is not restricted to Hadoop people anymore.

Best regards,

Simon

**Comment by wetneb** [comment]
We can totally add Parquet import support. Yes we are interested in good interoperability. I think the reason why @thadguidry asked those questions is that sometimes people come to us requesting "support for X" and it's just not clear if they are just looking at a feature list and want OpenRefine to tick one more box, or if that's a concrete need they have themselves. In the latter case, to make sure we actually address their need, it's worth understanding the broader picture (what do they want to do with the data once it's imported, for instance).

So I'd be curious to know the size (number of rows, columns, size on disk) of the file you are importing. (And even, if you'd be okay to share: where you got it from, what sorts of data it contains…). For instance, if the size of your Parquet file is way beyond the current scaling capabilities of OpenRefine 3, then it would mean that it's probably only worth implementing this for OpenRefine 4, because you wouldn't really be able to work on the file in OpenRefine 3.

**Comment by simonaubertbd** [comment]
@thadguidry  Yes, that's how I understood it. Sorry if my answer seemed rude, it wasn't on purpose.

@wetneb Hello. A good example of my usual files would be the Flights 1m dataset available here https://www.tablab.app/datasets/sample/parquet for an idea of the size. Sometimes, it can be bigger but usually, it's less than 20 millions (which is already big enough). I don't use it on very small files (less than 10 000 rows) seems it won't help.
if it's too big for OpenRefine 3, no urge, it can wait OpenRefine 4 but in the end, I'm pretty sure it should be on your roadmap ;)

Best regards,

Simon

**Comment by wetneb** [comment]
Ok, I'd say 1m rows is already a stretch for OpenRefine 3 (of course, given enough RAM, it can handle it, but it's not great UX), but it makes sense for OpenRefine 4.

**Comment by thadguidry** [comment]
Giving my 👍 that I need this now.  More and more scientific data is being placed in Parquet format.  The Parquet files themselves are often stored in HDF5/NetCDF4 directories in S3 or Google Cloud Store where HDF5 is the data container (like a TAR or PDF can contain files and metadata).  [OPeNDAP](https://www.opendap.org/support/OPeNDAP-clients) is used often in oceanography and other data collection/sensing fields where there are client libraries.

[Apache Parquet](https://parquet.apache.org/) is a good library and constantly maintained that we could utilize I think for a Parquet importer in OpenRefine 4.
Later, HDF5 support #640 can be added to OpenRefine eventually.

**Comment by jze** [comment]
Since nothing has happened with this ticket for so long, I assume that nobody is working on a Parquet import. I would then start working on the import.

**Comment by jze** [comment]
After spending a little more time on the the Apache Parquet format, I wonder if it is better to treat it parquet as a table (that's the only way I've ever encountered it) or as similar to a JSON file. It seems to be possible to store complex nested data structures within a parquet file. 

It would probably be best to have an option to treat it the some one or the other. However, this will makes implementation more complicated. But I guess that's the fun of it. 😉

**Comment by simonaubertbd** [comment]
@jze  Interesting. I would go at first with treating it like a table (should work for like 95% of cases).
